apstools/devices/motor_factory.py

```diff
@@ -0,0 +1,136 @@
+"""
+Motor Bundle Factory
+====================
+
+Factory for creating bundles with any number of positioners.
+
+.. autosummary::
+
+    ~axis_component
+    ~mb_class_factory
+    ~mb_creator
+
+.. rubric:: Examples
+
+.. code-block:: python
+    :linenos:
+
+    xy_stage = mb_creator(prefix="IOC:", motors={"x": "m21", "y": "m22"})
+
+.. seealso:: :doc:`How to use 'mb_creator()' </examples/ho_motor_factory>`
+"""
+
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Type
+from typing import Union
+
+from ophyd import Component
+from ophyd import EpicsMotor
+from ophyd import Kind
+from ophyd import MotorBundle
+from ophyd import SoftPositioner
+
+from apstools.utils import dynamic_import
+
+DEFAULT_DEVICE_BASE_CLASSES: List[Union[str, Type]] = [MotorBundle]
+DEFAULT_DEVICE_CLASS_NAME: str = "MB_Device"
+DEFAULT_MOTOR_LABELS: List[str] = ["motors"]
+DEFAULT_KIND: Kind = Kind.hinted | Kind.normal
+
+MOTORS_TYPE = Union[List[str], Dict[str, Optional[Union[str, Dict]]]]
+
+
+def axis_component(parms: dict[str, Any]) -> Component:
+    """Return a Component with 'parms for this axis."""
+    parms = parms or {}
+    if isinstance(parms, str):
+        parms = dict(prefix=parms)
+    attrs = dict(kind=DEFAULT_KIND, labels=DEFAULT_MOTOR_LABELS)
+    has_prefix: bool = parms.get("prefix") is not None
```

Comment:
I think this is mutating the input argument *parms*, right? Might be confusing if client code is not expecting it. Could *parms* be created anew (e.g. ``parms = {**parms}``)? Or else the doc string could make it clear the argument will be mutated?

Reply:
Good point: Don't mutate the input.  Define a local object from the input.  Apply throughout.  Lather, rinse, repeat.

Reply:
refactored to avoid mutating the code

Reply:
See c5d7cd6